lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,477 @@
+            }
+
+            if (_tasks.ContainsKey(taskId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already exists.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (taskConfiguration == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The task configuration is null."), Logger);
+            }
+
+            if (activeContext == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The context is null."), Logger);
+            }
+
+            if (NumberOfTasks >= _totalExpectedTasks)
+            {
+                string msg = string.Format("Trying to add an additional Task {0}, but the total expected Task number {1} has been reached.", taskId, _totalExpectedTasks);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _tasks.Add(taskId, new TaskInfo(new TaskStateMachine(), taskConfiguration, activeContext));
+
+            if (NumberOfTasks == _totalExpectedTasks && !MasterTaskExists())
+            {
+                Exceptions.Throw(new IMRUSystemException("There is no master task added."), Logger);
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of tasks in the task collection
+        /// </summary>
+        internal int NumberOfTasks
+        {
+            get { return _tasks.Count; }
+        }
+
+        /// <summary>
+        /// This method is called when receiving IRunningTask event during task submitting.
+        /// Adds the IRunningTask to the running tasks collection and update the task state to TaskRunning.
+        /// Throws IMRUSystemException if running tasks already contains this task or tasks collection doesn't contain this task.
+        /// </summary>
+        /// <param name="runningTask"></param>
+        internal void RecordRunningTask(IRunningTask runningTask)
+        {
+            if (_runningTasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already running.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (!_tasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _runningTasks.Add(runningTask.Id, runningTask);
+            UpdateState(runningTask.Id, TaskStateEvent.RunningTask);
+        }
+
+        /// <summary>
+        /// This method is called at the beginning of the recovery.
+        /// Clears the task collection, running task collection and resets the number of application error. 
+        /// </summary>
+        internal void Reset()
```

Comment:
Shouldn't _masterTaskId also be resetted? Infact to avoid any issues, I would remove this function and re-instantiate a new TaskManager class when we are ready again after faults.